backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/interest/InterestComputationService.java

```diff
@@ -0,0 +1,397 @@
+			@Nullable final FlatrateTermId onlyForContractId)
+	{
+		final ModularContractLogQuery query = ModularContractLogQuery.builder()
+				.computingMethodType(request.getComputingMethodType())
+				.flatrateTermId(onlyForContractId)
+				.invoicingGroupId(request.getInvoicingGroupId())
+				.billable(true)
+				.processed(false)
+				.lockOwner(request.getLockOwner())
+				.logEntryDocumentType(LogEntryDocumentType.SHIPPING_NOTIFICATION)
+				.orderBy(ModularContractLogQuery.OrderBy.TRANSACTION_DATE_ASC)
+				.build();
+
+		return modularContractLogService.streamModularContractLogEntries(query);
+	}
+
+	@NonNull
+	private Stream<ModularContractLogEntry> streamInterimInvoiceLogEntries(
+			@NonNull final FlatrateTermId contractId,
+			@NonNull final InterestComputationRequest request)
+	{
+		final ModularContractLogQuery query = ModularContractLogQuery.builder()
+				.computingMethodType(request.getComputingMethodType())
+				.flatrateTermId(contractId)
+				.invoicingGroupId(request.getInvoicingGroupId())
+				.billable(true)
+				.processed(false)
+				.onlyIfAmountIsSet(true)
+				.lockOwner(request.getLockOwner())
+				.logEntryDocumentType(LogEntryDocumentType.INTERIM_INVOICE)
+				.orderBy(ModularContractLogQuery.OrderBy.TRANSACTION_DATE_ASC)
+				.build();
+
+		return modularContractLogService.streamModularContractLogEntries(query);
+	}
+
+	private void distributeInterest(
+			@NonNull final InterestComputationRequest request,
+			@NonNull final BigDecimal interestDistributionPercent)
+	{
+		final FlatrateTermId onlyForContractId = null; // we want all contracts involved
+		streamShippingNotificationLogEntries(request, onlyForContractId)
+				.forEach(logEntry -> distributeInterest(request, interestDistributionPercent, logEntry));
+	}
+
+	private void distributeInterest(
+			@NonNull final InterestComputationRequest request,
+			@NonNull final BigDecimal interestDistributionPercent,
+			@NonNull final ModularContractLogEntry shippingNotification)
+	{
+		interestRepository.getForShippingNotificationLogId(request.getInterestRunId(), shippingNotification.getId())
+				.stream()
+				.map(interestLog -> interestLog.withComputedFinalInterest(interestDistributionPercent, request.getInterestToDistribute().getCurrencyId()))
+				.forEach(interestRepository::save);
+	}
+
+	@NonNull
+	private InvoiceAllocations initInvoiceAllocations(
+			@NonNull final InterestComputationRequest request,
+			final int additionalInterestDays,
+			@NonNull final ModularContractLogEntry invoice)
+	{
+		Check.assume(LogEntryDocumentType.INTERIM_INVOICE == invoice.getDocumentType(),
+					 "Expecting DocumentType = " + LogEntryDocumentType.INTERIM_INVOICE +
+							 " but got " + invoice.getDocumentType() + "! LogId=" + invoice.getId());
+		Check.assumeNotNull(invoice.getAmount(), "Invoices with no amount should've been skipped already! LogId="
+				+ invoice.getId());
+
+		final Instant conversionDate = Instant.now();
```

Comment:
pls avoid using Instant.now()!

pls introduce the run date as a field in InterestBonusComputationRequest.

More, you might want to refactor all code from this Service into some InterestBonusComputationCommand.